File: stdlib/public/core/ThreadLocalStorage.swift
//===--- ThreadLocalStorage.swift -----------------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//

import SwiftShims

// For testing purposes, a thread-safe counter to guarantee that destructors get
// called by pthread.
#if INTERNAL_CHECKS_ENABLED
internal class _TLSAtomicInt {
  internal var value: Int
  internal init() { self.value = 0 }

  internal var valuePtr: UnsafeMutablePointer<Int> {
    return unsafe _getUnsafePointerToStoredProperties(self).assumingMemoryBound(
      to: Int.self)
  }

  internal func increment() {
    _ = unsafe _swift_stdlib_atomicFetchAddInt(
      object: valuePtr,
      operand: 1)
  }

  internal func load() -> Int {
    return unsafe _swift_stdlib_atomicLoadInt(object: valuePtr)
  }
}

internal let _destroyTLSCounter = _TLSAtomicInt()

public // @testable
func _loadDestroyTLSCounter() -> Int {
  return _destroyTLSCounter.load()
}

#endif

// Thread local storage for all of the Swift standard library
//
// @moveonly/@pointeronly: shouldn't be used as a value, only through its
// pointer. Similarly, shouldn't be created, except by
// _initializeThreadLocalStorage.
//
internal struct _ThreadLocalStorage {
  // private: Should only be called by _initializeThreadLocalStorage
  internal init() {}

  // Get the current thread's TLS pointer. On first call for a given thread,
  // creates and initializes a new one.
  internal static func getPointer()
    -> UnsafeMutablePointer<_ThreadLocalStorage>
  {
    return unsafe _swift_stdlib_threadLocalStorageGet().assumingMemoryBound(
      to: _ThreadLocalStorage.self)
  }
}

// Destructor to register with pthreads. Responsible for deallocating any memory
// owned.
@_silgen_name("_stdlib_destroyTLS")
internal func _destroyTLS(_ ptr: UnsafeMutableRawPointer?) {
  unsafe _internalInvariant(ptr != nil,
    "_destroyTLS was called, but with nil...")
  let tlsPtr = unsafe ptr!.assumingMemoryBound(to: _ThreadLocalStorage.self)
  unsafe tlsPtr.deinitialize(count: 1)
  unsafe tlsPtr.deallocate()

#if INTERNAL_CHECKS_ENABLED
  // Log the fact we've destroyed our storage
  _destroyTLSCounter.increment()
#endif
}

@_silgen_name("_stdlib_createTLS")
internal func _createThreadLocalStorage()
  -> UnsafeMutablePointer<_ThreadLocalStorage>
{
  let tlsPtr: UnsafeMutablePointer<_ThreadLocalStorage>
    = UnsafeMutablePointer<_ThreadLocalStorage>.allocate(
      capacity: 1
  )
  unsafe tlsPtr.initialize(to: _ThreadLocalStorage())

  return unsafe tlsPtr
}

File: stdlib/public/stubs/ThreadLocalStorage.cpp
//===--- ThreadLocalStorage.cpp -------------------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//

#include <cstring>

#include "SwiftShims/ThreadLocalStorage.h"
#include "swift/Runtime/Debug.h"
#include "swift/Threading/ThreadLocalStorage.h"

SWIFT_CC(swift) SWIFT_RUNTIME_STDLIB_API
void _stdlib_destroyTLS(void *);

SWIFT_CC(swift) SWIFT_RUNTIME_STDLIB_API
void *_stdlib_createTLS(void);

SWIFT_RUNTIME_STDLIB_INTERNAL
void *
_swift_stdlib_threadLocalStorageGet(void) {

#if SWIFT_THREADING_NONE

  // If there's no threading, we can just keep a static variable.
  static void *value = _stdlib_createTLS();
  return value;

#elif SWIFT_THREADING_USE_RESERVED_TLS_KEYS

  // If we have reserved keys, use those
  void *value = swift::tls_get(swift::tls_key::stdlib);
  if (value)
    return value;

  static swift::once_t token;
  swift::tls_init_once(token, swift::tls_key::stdlib,
                       [](void *pointer) { _stdlib_destroyTLS(pointer); });

  value = _stdlib_createTLS();
  swift::tls_set(swift::tls_key::stdlib, value);
  return value;

#else // Threaded, but not using reserved keys

  // Register a key and use it
  static swift::tls_key_t key;
  static swift::once_t token;

  swift::tls_alloc_once(token, key,
                        [](void *pointer) { _stdlib_destroyTLS(pointer); });

  void *value = swift::tls_get(key);
  if (!value) {
    value = _stdlib_createTLS();
    swift::tls_set(key, value);
  }
  return value;

#endif
}
